Fix #39 Make the model classes immutable

ppalaga opened the pull request:
@angelozerr the code works the core tests are passing same as before. 

Could you please roughly review this and approve the general idea?

Once you approve, I'd add more JavaDoc to the changed classes.

I also adapted `ec4e` to these changes: https://github.com/ppalaga/ec4e/commits/ec4j33

angelozerr commented on core/src/main/java/org/eclipse/ec4j/core/model/Section.java:
```diff
@@ -17,148 +17,186 @@
+        public EditorConfig.Builder closeSection() {
+            if (glob == null) {
+                /* this is the first glob-less section */
+                Option rootOption = options.remove("root");
```
Please use OptionNames .root.name()

ppalaga replied:
> use OptionNames .root.name() instead of using "root"

Good catch, thanks!

angelozerr commented on core/src/main/java/org/eclipse/ec4j/core/parser/EditorConfigModelHandler.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Override
+    public void startPattern(ParseContext context) {
+        // TODO Auto-generated method stub
```
Remove this comments

ppalaga replied:
+1, thanks!

angelozerr commented:
> @angelozerr the code works the core tests are passing same as before.

Great! I must add too test for validation (services). I will do it after your PR.

> Could you please roughly review this and approve the general idea?

Love it! I have added few comments.

> Once you approve, I'd add more JavaDoc to the changed classes.

> I also adapted ec4e to these changes: https://github.com/ppalaga/ec4e/commits/ec4j33

I'm waiting for this ec4e PR to accept this ec4j PR

angelozerr commented on core/src/main/java/org/eclipse/ec4j/core/model/EditorConfig.java:
```diff
@@ -17,80 +17,171 @@
-    private String dirPath;
+        ResourcePath resourcePath;
+        Boolean root;
+        List<Section> sections = new ArrayList<>();
```
Should be better to initialize sections in the construcor and set sections a final, no?

ppalaga commented:
3eff4df replaces the previous head. 

In the fix for #39, I accepted both of your proposals and I added JavaDoc to the changed classes.

I added a few other fixes, that you please feel free to review/approve/veto each separately. If you find something too controversial, I can remove the given commit and submit it separately, so that we can properly discuss them.

angelozerr commented:
@ppalaga I'm waiting for ec4e PR to accept this PR. Thanks!

ppalaga commented:
> I'm waiting for ec4e PR to accept this PR.

Done: https://github.com/angelozerr/ec4e/pull/21

angelozerr commented:
Thanks a lot @ppalaga for your great contribution!
